### PR TITLE
pkg/storage/redis: add authentication support

### LINKED ISSUE #1157: pkg/storage/redis: Support Authentication
**Describe the solution you'd like**

Storage backend should support redis user/password authentication.

**Additional context**

Docs here https://redis.io/topics/acl

I think we can just support the newer user + password.

### PR DESCRIPTION
**Pull request by cuonglm**
## Summary
Add support for redis auth.

## Related issues
Fixes #1157

**Checklist**:
- [x] add related issues
- [x] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

**Comment by calebdoxsey** [comment]
@cuonglm would it be possible to use a data connection string instead? Something like:

`redis://username:password@host/db`

Redigo supports a `DialURL` method following this scheme: https://www.iana.org/assignments/uri-schemes/prov/redis.

This will make a future SQL provider easier to implement.

**Comment by cuonglm** [comment]
> @cuonglm would it be possible to use a data connection string instead? Something like:
> 
> `redis://username:password@host/db`
> 
> Redigo supports a `DialURL` method following this scheme: https://www.iana.org/assignments/uri-schemes/prov/redis.
> 
> This will make a future SQL provider easier to implement.

Yes, you can do it already without this CL. The `username/password` mechanism is for storage that generally don't have DSN for connection. With redis, it supports both way, so it's ok

**Comment by travisgroth** [comment]
> > @cuonglm would it be possible to use a data connection string instead? Something like:
> > `redis://username:password@host/db`
> > Redigo supports a `DialURL` method following this scheme: https://www.iana.org/assignments/uri-schemes/prov/redis.
> > This will make a future SQL provider easier to implement.
> 
> Yes, you can do it already without this CL. The `username/password` mechanism is for storage that generally don't have DSN for connection. With redis, it supports both way, so it's ok.

Hmm I actually tried a few combinations of URL with the current connection string option.  I think we need to use a different Dial() to support that.  But maybe I missed something?

> databroker_storage_connection_string: redis:http://pomerium:badpassword@127.0.0.1:6379/0

> databroker_storage_connection_string: redis:http://pomerium:badpassword@127.0.0.1:6379

> databroker_storage_connection_string: redis://pomerium:badpassword@127.0.0.1:6379/0

> databroker_storage_connection_string: redis://pomerium:badpassword@127.0.0.1:6379

I believe all of these gave me `too many colons in address`

**Comment by calebdoxsey** [comment]
In my opinion if a DSN is able to use usernames and passwords for redis, we should have only one way of configuring redis.

**Comment by cuonglm** [comment]
> In my opinion if a DSN is able to use usernames and passwords for redis, we should have only one way of configuring redis.

just concern if we want to support other storage which does not have DSN support. @travisgroth how do you think?

**Comment by cuonglm** [comment]
> In my opinion if a DSN is able to use usernames and passwords for redis, we should have only one way of configuring redis.

fixed.

File: docs/configuration/readme.md [hunk 1]
```diff
@@ -831,7 +831,7 @@ The backend storage that databroker server will use, available types: `memory`,
 - Config File Key: `databroker_storage_connection_string`
 - Type: `string`
 - **Required** when storage type is `redis`
-- Example: `":6379"`
+- Example: `"redis://localhost/6379/0"`
```
Comment (calebdoxsey) [inline docs/configuration/readme.md:834]: this should be `redis://localhost:6379/0`

Comment (cuonglm) [reply docs/configuration/readme.md]: Fixed

**Pull request merged**